### PR TITLE
Replaced Grafana graph with Chartkick (highcharts)

### PR DESCRIPTION
**Pull request by Nakushi64**
Screenshot 2021-02-19 at 17 24 57

Fixes https://github.com/Safecast/safecastapi/issues/776

### CODE REVIEW

File: app/models/ingest_measurement.rb [hunk 1]
```diff
@@ -11,5 +11,117 @@ class << self
...
+                           "gte": 'now-8w',
+                           "lte": 'now',
+                           "format": 'epoch_millis'
+                         }
+                       }
+                     },
+                     {
+                       "query_string": {
+                         "analyze_wildcard": true,
+                         "query": "device_urn:" + "\"" + device_urn + "\"" # rubocop:disable all
```
Comment (matschaffer) [inline app/models/ingest_measurement.rb:31]: You can make this cleaner with a [term query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html)

It should just be
```ruby
{
  "term": {
    "device_urn": device_urn
  }
}
```

File: app/models/ingest_measurement.rb [hunk 2]
```diff
@@ -11,5 +11,117 @@ class << self
...
+                 }
+               }
+             }
+    end
+
+    def query_last_sensor_location(device_urn)
+      search "query": {
+        "match": { "device_urn": device_urn }
+      },
+             "size": 1,
```
Comment (matschaffer) [inline app/models/ingest_measurement.rb:123]: Indentation seems weird here. Might be nice to clean up as long as you can get rubycop happy.

File: app/models/ingest_measurement.rb [hunk 3]
```diff
@@ -11,5 +11,117 @@ class << self
...
+    def query_sensor_data(device_urn) # rubocop:disable Metrics/MethodLength
+      search "query":
+                 { "bool": {
+                   "filter": [
+                     {
+                       "range": {
+                         "@timestamp": {
+                           "gte": 'now-8w',
+                           "lte": 'now',
+                           "format": 'epoch_millis'
```
Comment (matschaffer) [inline app/models/ingest_measurement.rb:24]: Don't think you need this line.

**Review by matschaffer — COMMENTED** [review]
Looks good overall. Some comments. Is this "ready to go" for you or is there stuff you still want to add?

**Comment by matschaffer** [comment]
Also I think we can add stuff like mapkick.js to code climate exclusions https://docs.codeclimate.com/docs/excluding-files-and-folders#excluding-specific-files-and-file-types-at-any-level

**Comment by Nakushi64** [comment]
@matschaffer I made all the changes you recommended and added more graphs. We can give it another look on the dev server if you want.

**Comment by Nakushi64** [comment]
@matschaffer looks like enabling Rubocop for Codeclimate did not work.

**Comment by eitoball** [comment]
@Nakushi64 

Please change `.codeclimate.yml` as following and try.

```diff
diff --git a/.codeclimate.yml b/.codeclimate.yml
index b8861e87..2e67c587 100644
--- a/.codeclimate.yml
+++ b/.codeclimate.yml
@@ -1,6 +1,7 @@
 plugins:
   rubocop:
     enabled: true
+    channel: rubocop-0-92
 exclude_patterns:
 - "config/"
 - "db/"
 ```

**Comment by matschaffer** [comment]
Looks like still codeclimate issues

**Comment by Nakushi64** [comment]
Only 1 issue left. Will tomorrow try to solve it as well.

File: app/models/ingest_measurement.rb [hunk 4]
```diff
@@ -11,5 +11,129 @@ class << self
...
+               "sensor_data": {
+                 "date_histogram": {
+                   "interval": '12h',
+                   "field": '@timestamp',
+                   "min_doc_count": 0,
+                   "extended_bounds": {
+                     "min": 'now-8w',
+                     "max": 'now'
+                   },
+                   "format": 'epoch_millis'
```
Comment (eitoball) [inline app/models/ingest_measurement.rb:44]: If you use this `format` instead,

```suggestion
                   "format": 'yyyy-MM-dd HH'
```

you can use `bucket['key_as_string']` instead on line 97 of `app/helpers/device_stories_helper.rb`.

File: app/helpers/device_stories_helper.rb [hunk 5]
```diff
@@ -72,4 +72,45 @@ def last_air_quality_values(last_values)
...
+  def sensor_names
+    { 'radiation_sensors' => %w(lnd_7128ec lnd_7318c lnd_712u lnd_7318u lnd_78017w lnd7318u lnd7128c),
+      'air_sensors' => %w(pms_pm10_0 pms_pm02_5 pms_pm01_0),
+      'bat_voltage' => ['bat_voltage'],
+      'temperature_C' => ['temperature_C'],
+      'temperature_F' => ['temperature_F'],
+      'humidity' => ['humidity'],
+      'pressure' => ['pressure'],
+      'charging' => ['charging'] }
+  end
```
Comment (eitoball) [inline app/helpers/device_stories_helper.rb:86]: This should be a constant. We don't want to create this structure every time.

File: app/helpers/device_stories_helper.rb [hunk 6]
```diff
@@ -72,4 +72,45 @@ def last_air_quality_values(last_values)
...
+          sensor_value = bucket[sensor]['value']
+          if sensor_value then sensor_exists = true end
+          hash_sensor.merge!(date => sensor_value)
+        end
+        if sensor_exists then sensor_type_hashes.push({ "name": sensor, "data": hash_sensor }) end
+      end
+      all_hashes.merge!(sensor_names.key(sensor_type) => sensor_type_hashes)
+    end
+    all_hashes
+  end
```
[the remark this answers is not in the document]
Comment (Nakushi64) [reply app/helpers/device_stories_helper.rb:107]: @eitoball Looks very elegant. Thanks for helping me out.

File: app/models/ingest_measurement.rb [hunk 7]
```diff
@@ -11,5 +11,129 @@ class << self
...
+                       "term": {
+                         "device_urn": device_urn
+                       }
+                     }
+                   ]
+                 } },
+             "aggs": {
+               "sensor_data": {
+                 "date_histogram": {
+                   "interval": '12h',
```
Comment (eitoball) [inline app/models/ingest_measurement.rb:37]: It seems that we should use `fixed_interval` instead.

```suggestion
                   "fixed_interval": '12h',
```

I got this message from elasticsearch.

```
[interval] on [date_histogram] is deprecated, use [fixed_interval] or [calendar_interval] in the future.
```

**Comment by matschaffer** [comment]
Talking with @nokton today and we'll just go ahead with this and worry about making codeclimate happy later

**Review by robouden — COMMENTED** [review]
Seems fine to me.

**Pull request merged**